Complete Databricks exporter and mixin package

aalhour opened the pull request:
Fixes: https://github.com/grafana/cloud-onboarding/issues/9896
Fixes: https://github.com/grafana/cloud-onboarding/issues/9914
Part of: https://github.com/grafana/cloud-onboarding/issues/9716

### Dashboards

#### Overview Dashboard

<img width="2531" height="1204" alt="Overview" src="https://github.com/user-attachments/assets/a0dab06f-3eb6-49ce-b3db-3a138a54fc44" />

#### Jobs & Pipelines Dashboard

<img width="2560" height="1269" alt="Jobs   Pipelines" src="https://github.com/user-attachments/assets/dcabb77a-be76-447c-8749-606f5f6a9ce3" />

#### Warehouses & Queries Dashboard

<img width="2560" height="1294" alt="Warehouse" src="https://github.com/user-attachments/assets/6851b7ec-026b-4552-9848-75f1b133db78" />

Dasomeone commented on mixin/README.md:
```diff
@@ -0,0 +1,241 @@
+# Databricks Mixin
+
+A set of Grafana dashboards and Prometheus alerts for monitoring Databricks.
```
Drop `prometheus` here as they're just alerts :)
Alternatively call them prometheus-compatible

Dasomeone commented on mixin/g.libsonnet:
```diff
@@ -0,0 +1,5 @@
+// Auto-generated file, do not edit
+local grafonnet = import 'github.com/grafana/grafonnet/gen/grafonnet-v11.4.0/main.libsonnet';
+
+grafonnet
```
Unnecessary

Dasomeone commented on mixin/signals/jobs_and_pipelines.libsonnet:
```diff
@@ -0,0 +1,371 @@
+      nameShort: 'Job status',
+      description: 'Job status counts by job name (adapts to dashboard time range).',
+      type: 'raw',
+      unit: 'none',  // would be 'runs' but no custom unit available
```
Instead of `none` use `short` as a basic number

Dasomeone commented on mixin/panels.libsonnet:
```diff
@@ -0,0 +1,766 @@
+        g.panel.gauge.new('Jobs success %')
+        + g.panel.gauge.queryOptions.withTargets([signals.jobsAndPipelines.jobSuccessRate.asTarget()])
+        + g.panel.gauge.panelOptions.withDescription('Percentage of successful jobs (adapts to dashboard time range).')
+        + g.panel.gauge.standardOptions.withUnit('percentunit')
```
Nit: Is it necessary to override the signal unit here, or does it not propagate properly

Dasomeone commented on mixin/panels.libsonnet:
```diff
@@ -0,0 +1,766 @@
+        )
+        + g.panel.stat.options.withGraphMode('none')
+        + g.panel.stat.options.withTextMode('value')
+        + g.panel.stat.standardOptions.withUnit('none'),  // would be 'dbu' but no custom unit available
```
You could create a custom unit? https://github.com/grafana/grafonnet/blob/8a27651b56fbdba45a389ccb11440200091c73a1/docs/API/panel/stat/index.md#fn-standardoptionswithunit

Dasomeone replied:
This shouldn't have been resolved? I don't see any changes to the mixin? 
<img width="202" height="83" alt="image" src="https://github.com/user-attachments/assets/1158b248-0aeb-4141-bb7f-f6099dd07e9a" />

aalhour commented on mixin/signals/overview.libsonnet:
```diff
@@ -0,0 +1,156 @@
+      unit: 'currencyUSD',
+      sources: {
+        prometheus: {
+          expr: 'sum by (workspace_id) (databricks_billing_cost_estimate_usd{%(queriesSelector)s})',
```
I've given this a thought and it reminded me of the [Prometheus naming convension](https://prometheus.io/docs/practices/naming/), wouldn't this suggestion break the convention? I have seen metrics with suffixes like `_24h` in the past but these always came directly from the source system. But in databricks they don't. We're building our custom exporter like that? I'm split on this issue 🤔

Dasomeone replied:
Commented on this in the exporter part as well, I think it makes sense. `_total` is generally reserved as for being actual totals and having time slice windows would go against that. For the others where it's clear that it's a configurable window, we should have that exposed. If not in the name, then certainly in the HELP description

aalhour replied:
I replied here: https://github.com/grafana/databricks-prometheus-exporter/pull/14#discussion_r2626827284. Let's discuss it there.

Dasomeone commented on cmd/databricks-exporter/main.go:
```diff
@@ -34,14 +34,12 @@ import (
 )
 
 var (
```
Potential: Add flags/config options for configurable variables, e.g. timeouts, window sizes, etc?

Dasomeone commented on collector/collector.go:
```diff
@@ -47,115 +61,119 @@ func openDatabricksDatabase(config *Config) (*sql.DB, error) {
+	db := sql.OpenDB(connector)
+
+	// Configure connection pool for better resilience
+	db.SetMaxOpenConns(10)                 // Limit concurrent connections to avoid overwhelming Databricks
```
Should probably set some timeouts too?

Dasomeone commented on collector/collector.go:
```diff
@@ -28,11 +30,23 @@ import (
 	labelWorkspaceID = "workspace_id"
 	labelSKUName     = "sku_name"
 	labelCloud       = "cloud"
 	labelUsageUnit   = "usage_unit"
```
+1 only in tests

Dasomeone commented on collector/metrics.go:
```diff
@@ -0,0 +1,241 @@
+
+		// ===== Pipelines Metrics (SRE/Platform) =====
+
+		PipelineRunsTotal: prometheus.NewDesc(
```
Same as we discussed for `job_name` and `job_id`. Would be good to have a conditional here to let customers pick which to include, and perhaps default to only one of the two?

aalhour replied:
As discussed on Slack, I opted in to keep both `job_id` and `job_name` since they have a 1:1 mapping, same goes for `pipeline_id` and `pipeline_name`. However, you were right to call out `task_id`, it definitely has a high cardinality since a single job can have many tasks and the mapping is 1:M.

(I also replied to your question about our Slack convo in the other comment about this.)

Dasomeone commented on collector/queries.go:
```diff
@@ -0,0 +1,415 @@
+	`
+
+	// jobSLAMissQuery identifies jobs exceeding a configured SLA threshold.
+	// Currently hardcoded to 3600 seconds (1 hour) - should be configurable.
```
If we want it to be configurable, then please provide it as a default value in a variable, ideally one that you can configure in a command-line flag or config file :)

Dasomeone commented on collector/queries.go:
```diff
@@ -0,0 +1,415 @@
+	// pipelineFreshnessLagQuery calculates data freshness lag per pipeline.
+	// TODO: Implement based on actual pipeline metadata and target watermarks.
+	// This requires understanding the specific pipeline freshness requirements.
+	// Placeholder query - needs refinement based on actual use case.
```
This is a placeholder query, let's not keep it in if we don't want it as-is

Dasomeone commented on collector/queries.go:
```diff
@@ -0,0 +1,415 @@
+)
+
+// Time window constants for configurable lookback periods
+const (
```
Would be great to use these values directly in the sql queries, eg. the one right above with `INTERVAL 1 HOUR` and another that I commented on with `3600` seconds

Dasomeone commented on collector/collector.go:
```diff
@@ -47,115 +61,119 @@ func openDatabricksDatabase(config *Config) (*sql.DB, error) {
 	level.Debug(c.logger).Log("msg", "Collecting metrics.")
 
-	var up float64 = 1
 	// Open a new connection to the database each time; This makes the connection more robust to transient failures
```
Thought about this some more, I think this is something we need to do. 
Databricks may end up seeing it as spammy connections and rate limit us, and it also is just less expensive to run and execute, and stays more closely aligned with the TLS contract around reuse and long-running connections

Dasomeone commented on collector/queries.go:
```diff
@@ -0,0 +1,822 @@
+	`
+)
+
+// Time window constants for configurable lookback periods
```
Only used in tests, testing that they exist

Dasomeone commented on collector/queries_test.go:
```diff
@@ -0,0 +1,295 @@
+	}
+}
+
+func TestTimeWindowConstants(t *testing.T) {
```
If anything, you could test that the default constants in config.go are configured properly
```go
const (
DefaultQueryTimeout        = 5 * time.Minute
DefaultBillingLookback     = 24 * time.Hour
DefaultJobsLookback        = 2 * time.Hour
DefaultPipelinesLookback   = 2 * time.Hour
DefaultQueriesLookback     = 1 * time.Hour
DefaultSLAThresholdSeconds = SecondsPerHour
)
```

Dasomeone commented on README.md:
```diff
@@ -112,34 +135,239 @@ The Service Principal needs:
+
+- **Source table:** `system.billing.usage`
+- **Type:** Counter
+- **Labels:** `workspace_id`, `sku_name`, `usage_date`
```
I don't see `usage_date` added as a label anywhere in the codebase unless I'm missing something?

aalhour replied:
I will address this point as part of improving the README.

Dasomeone commented on README.md:
```diff
@@ -112,34 +135,239 @@ The Service Principal needs:
+
+These metrics help with FinOps and cost tracking.
+
+#### `databricks_billing_dbus_total`
```
Suggest we be very explicit that this is a sliding window, and not an actual total. 
We should probably rename any of the metrics with sliding windows to have their default value in the name?

Dasomeone commented on mixin/README.md:
```diff
@@ -0,0 +1,240 @@
+- ✅ Skips collection if unavailable
+- ✅ Rechecks periodically (every 10 scrapes)
+- ✅ Automatically resumes when permissions are fixed
+- ✅ Continues collecting all other metrics normally
```
Personal preference, I wouldn't include this, certainly not in the mixin readme

aalhour commented on collector/metrics.go:
```diff
@@ -0,0 +1,241 @@
+
+		// ===== Jobs Metrics (SRE/Platform) =====
+
+		JobRunsTotal: prometheus.NewDesc(
```
That's correct, the labels `job_id` and `job_name` have 1:1 mapping, same thing for `pipeline_id` and `pipeline_name`, but `task_id` which refers to smaller steps in jobs/pipelines has a 1:M mappings with jobs and pipelines, that's why I added a flag having it disabled by default (not collected unless explicitly enabled by the customer).

aalhour commented on collector/metrics.go:
```diff
@@ -0,0 +1,241 @@
+
+		// ===== Exporter Health =====
+
+		Up: prometheus.NewDesc(
```
I agree, first I will rename the metric and then work on proposing a metamonitoring metric for tracking "data flowing in" as expected.

aalhour commented on collector/collector.go:
```diff
@@ -47,115 +61,119 @@ func openDatabricksDatabase(config *Config) (*sql.DB, error) {
-	level.Debug(c.logger).Log("msg", "Finished collecting billing metrics.")
-	return rows.Err()
+	// Emit up=1 immediately after successful connection
+	// This ensures the metric is always emitted even if subsequent collection hangs/times out
```
That's a good point, I will thinking of an additional metric after I rename the up metric to be about the exporter itself.

aalhour commented on collector/billing.go:
```diff
@@ -0,0 +1,199 @@
+type BillingCollector struct {
+	db      *sql.DB
+	metrics *MetricDescriptors
+	logger  log.Logger
```
I'd like to use a single logger, this might have slipped through the cracks, apologies. I will address it as part of the renovate-bot suggestions, more on that here as to why these are connected: https://github.com/grafana/databricks-prometheus-exporter/pull/14#discussion_r2626539675

Dasomeone replied:
Fantastic, thanks!

aalhour commented on collector/billing_test.go:
```diff
@@ -0,0 +1,403 @@
+
+	// Collect all metrics - should continue despite first failure
+	ch := make(chan prometheus.Metric, 20)
+	err = collector.Collect(ch)
```
This is super weird :D a classic case of "it works on my machine"

```
$ go test ./collector
ok  	github.com/grafana/databricks-prometheus-exporter/collector	0.242s
```

aalhour commented on README.md:
```diff
@@ -112,34 +135,239 @@ The Service Principal needs:
 3. The **Application ID** is your **Client ID**
 4. Generate and copy the **Client Secret**
 
+## System Tables
```
Done.

Dasomeone replied:
It's not? Maybe you're missing a stash to push or something, metrics are still listed int he base readme.

I suggest keeping the base readme focused on what the exporter is and what pre-requisites there are to build it and configure Databricks, e.g. the permissions you call out on the system table doc.

The system table doc (and docs folder) is a good start, just need a second one for the metrics and then linking to them from the readme!

aalhour replied:
Done, changes incoming.

aalhour commented on mixin/README.md:
```diff
@@ -0,0 +1,240 @@
+
+## Alerts
+
+The mixin includes 16 alerts across three personas:
```
Correct, I need to stop doing this. I removed them.

aalhour commented on collector/queries.go:
```diff
@@ -0,0 +1,822 @@
+				AND COALESCE(q1.compute.warehouse_id, 'unknown') = COALESCE(q2.compute.warehouse_id, 'unknown')
+				AND q2.start_time <= q1.start_time
+				AND (q2.end_time >= q1.start_time OR q2.end_time IS NULL)
+			WHERE q1.start_time >= current_timestamp() - INTERVAL 1 HOUR
```
Good call out, I will add support for them in the utility func and add tests for it. Both formats are supported from Databricks, which I will add a comment for too.

aalhour commented on Makefile:
```diff
@@ -5,6 +5,20 @@ ALL_SRC := $(shell find . -name '*.go' -o -name 'Dockerfile*' -type f | sort)
+	@echo "Running integration benchmarks..."
+	@go test -tags=integration -v -bench=. -benchmem -timeout 10m ./collector/...
 
 include Makefile.common
```
Now it works.

```bash
$ make clean build

>> cleaning build artifacts
rm -rf bin
rm -f coverage.out coverage.html
>> building databricks-exporter
go build  -o bin/databricks-exporter ./cmd/databricks-exporter/
```

aalhour commented on README.md:
```diff
@@ -112,34 +135,239 @@ The Service Principal needs:
+- **Type:** Gauge
+- **Values:**
+  - `1` - Connection successful, metrics collected
+  - `0` - Connection failed or unable to collect metrics
```
This is done, I offloaded details about the system tables to a new doc under `docs/` and updated the contents based on the recent changes to the codebase.

aalhour commented on collector/collector.go:
```diff
@@ -82,7 +70,7 @@ func openDatabricksDatabase(config *Config) (*sql.DB, error) {
 type Collector struct {
 	config       *Config
-	logger       log.Logger
+	logger       *slog.Logger
```
Same, `promslog` configures a new `*slog.Logger` under the hood and returns it, this is the same logger.

aalhour commented:
Summary of updates:

- **Logging**
  - Migrated from `promlog/go-kit` to `slog/promslog` (Go stdlib)
  - Removed `github.com/go-kit/log` dependency

- **Metrics**
  - Renamed `databricks_up` → `databricks_exporter_up`
  - Renamed `databricks_billing_export_errors_total` → `databricks_billing_scrape_errors` (now a gauge)
  - Added `databricks_scrape_status{query}` per-collector health metric
  - Added `databricks_exporter_info{version,...}` for config visibility
  - Updated all HELP texts with lookback window info

- **Query builders**
  - Removed unused legacy query constants
  - Added `--table-check-interval` flag for optional table availability checks
  - Fixed `durationToSQLInterval` singular/plural grammar (`1 HOUR` vs `2 HOURS`) + tests

- **Performance**
  - Moved `CollectTaskRetries` check before query execution to avoid wasted DB calls

- **Dependencies**
  - Pinned all renovate-bot dependencies now that the exporter uses promslog

- **Tests**
  - Added tests for `durationToSQLInterval` and `Build*` functions
  - Removed zero-value tests (`TestLabelConstants`, `TestConstants`)
  - Updated mocks for `CollectTaskRetries=false` default

- **Docs**
  - Fixed metric types (Counter → Gauge for sliding windows)
  - Added all CLI flags and environment variables
  - Applied sentence case per Grafana style guide

Dasomeone commented on collector/queries_test.go:
```diff
@@ -1,198 +1,399 @@
-		{"queryErrorsQuery", queryErrorsQuery},
-		{"queryDurationQuery", queryDurationQuery},
-		{"queriesRunningQuery", queriesRunningQuery},
+		// Days - singular and plural
```
You mention in the function description on the actual implementation that:
24 = 1 DAY
48h = 2 DAYS

Could you add tests for this too?

Similarly are we also needing to account for Months/years, etc?

Dasomeone reviewed (commented):
Couple minor comments and some existing threads had separate comments added. I'm 99.9% happy with the exporter, fantastic work @aalhour !
